nae-object-library/src/main/java/com/netgrif/application/engine/objects/auth/domain/AbstractUser.java

```diff
@@ -1,106 +1,192 @@
-        if (negativeProcessRoles.stream().anyMatch(it -> it.getStringId().equals(role.getStringId())))
-            return;
-        negativeProcessRoles.add(role);
-    }
+    /** User's middle name (optional) */
+    protected String middleName;
 
-    public void removeNegativeProcessRole(ProcessRole role) {
-        negativeProcessRoles.remove(role);
+    /** User's last/family name */
+    @NotNull
+    protected String lastName;
+
+    /** User's email address */
+    protected String email;
+
+    /** URL or identifier of user's avatar image */
+    protected String avatar;
+
+    /**
+     * Constructs a new user with Object ID.
+     *
+     * @param id MongoDB ObjectId of the user
+     * @param realmId Security realm identifier
+     * @param username User's login name
+     * @param firstName User's first name
+     * @param middleName User's middle name
+     * @param lastName User's last name
+     * @param email User's email address
+     * @param avatar User's avatar URL/identifier
+     */
+    public AbstractUser(ObjectId id, String realmId, String username, String firstName, String middleName, String lastName, String email, String avatar) {
+        super(id, realmId);
+        this.username = username;
+        this.firstName = firstName;
+        this.middleName = middleName;
+        this.lastName = lastName;
+        this.email = email;
+        this.avatar = avatar;
     }
 
-    public void addGroup(Group group) {
-        this.groupIds.add(group.getStringId());
-        this.groups.add(group);
+    /**
+     * Constructs a new user with String ID.
+     *
+     * @param id String representation of MongoDB ObjectId
+     * @param realmId Security realm identifier
+     * @param username User's login name
+     * @param firstName User's first name
+     * @param middleName User's middle name
+     * @param lastName User's last name
+     * @param email User's email address
+     * @param avatar User's avatar URL/identifier
+     */
+    public AbstractUser(String id, String realmId, String username, String firstName, String middleName, String lastName, String email, String avatar) {
+        super(id, realmId);
+        this.username = username;
+        this.firstName = firstName;
+        this.middleName = middleName;
+        this.lastName = lastName;
+        this.email = email;
+        this.avatar = avatar;
     }
 
-    @Override
-    public void removeGroupId(String groupId) {
-        this.groupIds.remove(groupId);
+    /**
+     * Gets the user's password.
+     * @return the user's password
+     */
+    public abstract String getPassword();
+
+    /**
+     * Sets the user's password.
+     * @param password new password to set
+     */
+    public abstract void setPassword(String password);
+
+    /**
+     * Sets a credential for the user.
+     * @param key credential identifier
+     * @param credential credential value object
+     */
+    public void setCredential(String key, Credential<?> credential) {}
```

Comment:
Why are the following methods not abstract and have rather empty implementations?